### PR TITLE
Add support for nested types as non-key columns on joins 

### PR DESCRIPTION
**Pull request by nartal1**
This fixes https://github.com/NVIDIA/spark-rapids/issues/3116. 
Tested various joins(BHJ, SMJ, Shuffle Hash Join, Broadcast Nested loop join and Catersian Product) on maps, structs and arrays.
We already had support to most of them. This PR adds support for missing types and adds missing tests.

### CODE REVIEW

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 1]
```diff
@@ -3366,8 +3366,9 @@ object GpuOverrides extends Logging {
       (join, conf, p, r) => new GpuBroadcastNestedLoopJoinMeta(join, conf, p, r)),
     exec[CartesianProductExec](
       "Implementation of join using brute force",
-      ExecChecks(TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64 +
-          TypeSig.ARRAY.nested(TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64),
+      ExecChecks((TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64 +
+          TypeSig.ARRAY + TypeSig.MAP + TypeSig.STRUCT)
+          .nested(TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL_64),
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3371]: Why not support nested arrays and structs?  It's just a gather, and we support gather on arbitrarily nested structures.

File: integration_tests/src/main/python/join_test.py [hunk 2]
```diff
@@ -182,6 +182,16 @@ def do_join(spark):
         return left.join(right, left.key == right.r_key, join_type)
     assert_gpu_and_cpu_are_equal_collect(do_join, conf=_hash_join_conf)
 
+@validate_execs_in_gpu_plan('GpuShuffledHashJoinExec')
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen', [all_basic_struct_gen], ids=idfn)
+@pytest.mark.parametrize('join_type', all_join_types, ids=idfn)
+def test_hash_join_struct(data_gen, join_type):
```
Comment (jlowe) [inline integration_tests/src/main/python/join_test.py:189]: Shouldn't we simply combine `test_hash_join_array`, `test_hash_join_map`, and `test_hash_join_struct` into a single `test_hash_join` that is parameterized with all the various types to test?  That matches the approach we use almost everywhere else in the tests and reduces the amount of boilerplate for this type of test.

File: integration_tests/src/main/python/join_test.py [hunk 3]
```diff
@@ -255,6 +265,19 @@ def do_join(spark):
     conf = copy_and_update(allow_negative_scale_of_decimal_conf, {'spark.rapids.sql.batchSizeBytes': batch_size})
     assert_gpu_and_cpu_are_equal_collect(do_join, conf=conf)
 
+# local sort because of https://github.com/NVIDIA/spark-rapids/issues/84
+# After 3.1.0 is the min spark version we can drop this
+@ignore_order(local=True)
+@pytest.mark.order(1) # at the head of xdist worker queue if pytest-order is installed
+@pytest.mark.parametrize('data_gen', single_level_array_gens + map_string_string_gen + [all_basic_struct_gen], ids=idfn)
+@pytest.mark.parametrize('batch_size', ['100', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
+def test_cartesian_join_nestedtype_nonkey(data_gen, batch_size):
```
Comment (jlowe) [inline integration_tests/src/main/python/join_test.py:274]: Cartesian joins don't have a key, it's just a nested for loop to generate every row combination between the two tables.  This test should be combined with the previous test to test against all types, nested or not.

File: integration_tests/src/main/python/join_test.py [hunk 4]
```diff
@@ -314,6 +337,18 @@ def do_join(spark):
     conf = copy_and_update(allow_negative_scale_of_decimal_conf, {'spark.rapids.sql.batchSizeBytes': batch_size})
     assert_gpu_and_cpu_are_equal_collect(do_join, conf=conf)
 
+# local sort because of https://github.com/NVIDIA/spark-rapids/issues/84
+# After 3.1.0 is the min spark version we can drop this
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen', single_level_array_gens + map_string_string_gen + [all_basic_struct_gen], ids=idfn)
+@pytest.mark.parametrize('batch_size', ['100', '1g'], ids=idfn) # set the batch size so we can test multiple stream batches
+def test_broadcast_nested_loop_join_nestedtype_nonkey(data_gen, batch_size):
```
Comment (jlowe) [inline integration_tests/src/main/python/join_test.py:345]: Similar comment as the previous.  There's no "key" here,  it's a cross join.

File: integration_tests/src/main/python/join_test.py (hunk 2 above)
Comment (nartal1) [reply integration_tests/src/main/python/join_test.py]: Thanks for the review. You are right, combining nested types test reduces the boilerplate code. I have combined other tests as well(SMJ, BHJ). Please take another look.

File: integration_tests/src/main/python/join_test.py (hunk 3 above)
Comment (nartal1) [reply integration_tests/src/main/python/join_test.py]: combined with previous test

File: integration_tests/src/main/python/join_test.py (hunk 4 above)
Comment (nartal1) [reply integration_tests/src/main/python/join_test.py]: done.

**Review by jlowe — APPROVED** [review]

**Comment by nartal1** [comment]
build

**Pull request merged**